### ISSUE TITLE
Allow the tigera-operator Helm chart to change where the Calico resources are created

### ISSUE THREAD

**Issue by michelzanini** [projectcalico/calico#5876]
Hi,

When using the `tigera-operator` Helm chart it creates some namespaces, like:

* calico-system
* tigera-operator

We would like to keep those in `kube-system`. I have seen that there was a PR merged that allows to switch the `tigera-operator` namespace to another of the user's choice.

Is it possible to change the `calico-system` as well? I see some resources inside that namespace can be customised via a spec configuration in values.yaml.

I cant find an option for namespace name on that installation spec. Is it there somewhere? If not, can you add it?

Thanks.
Michel.

**Comment by caseydavenport** [comment]
Could you explain why you need the Calico components to exist in kube-system?

In general that would be bad practice in terms of security boundaries and separation of concerns, which is why we don't allow that today. We would need a compelling use-case for housing these in kube-system to make that configurable.

**Comment by caseydavenport** [comment]
> I have seen that there was a PR merged that allows to switch the tigera-operator namespace to another of the user's choice.

For context, the intention of that PR was not to enable users to run the operator in an alternative namespace (e.g., kube-system) although it was a side effect of it. 

The intention was to fix the helm chart so that the chart release was installed within the same namespace that the operator was installed in (prior it was always installed in the default namespace)

**Comment by michelzanini** [comment]
We try to limit the add-ons to be installed in a single namespace as we ofter a platform to developers, where they use namespaces for the environments. Using more namespaces make things more messy for them as it becomes noisy. The `kube-system` is a special namespace that is often hidden away on UI tools and Logging platforms, which is ideal for who is building a platform and want to hide those system components away.

Besides, before `tigera-operator`, was Calico not already using as default the `kube-system` namespace? What changed since then? Cilium often has examples installing it on `kube-system`.

Can you elaborate on why is less secure, as I don't seem to see any difference... Also, not asking to change the defaults, just to allow people to control it. Not everyone will want to run things on the same location. All Helm charts I have ever seen allow namespace customisation.

**Comment by caseydavenport** [comment]
> The kube-system is a special namespace

Yes, exactly, and Calico isn't part of Kubernetes - it **shouldn't** be in the Kubernetes system namespace.

>  that is often hidden away on UI tools and Logging platforms

In my opinion, a more appropriate fix for this would be to enhance those tools and platforms so that they can hide other namespaces beyond just kube-system.

> Besides, before tigera-operator, was Calico not already using as default the kube-system namespace?

Yep, it was, and we intentionally moved away from this deployment model in the operator for better separation of concerns.

> Cilium often has examples installing it on kube-system.

I don't maintain Cilium, so can't comment on why they feel that's OK. But if I were them, I would likely remove it from kube-system.

> Can you elaborate on why is less secure, as I don't seem to see any difference... 

It's not inherently a security flaw, however separating disjoint systems via namespaces is a best-practice that allows for a better security posture since namespaces are a natural isolation boundary. To name a few of the benefits:

- You can more easily use RBAC to limit access to resources so only those users who require access have it.
- You can use network policy more confidently without accidentally selecting something you didn't mean to. 
- You can prevent accidental configuration errors by more tightly scoping commands to a single application's "workspace".
- You can more easily debug and understand which resources belong to which applications.

There's really no reason Calico should run in the same namespace as the Kubernetes control plane - the two components don't need to share information, API resources / config, etc.

> All Helm charts I have ever seen allow namespace customisation.

That doesn't mean that it's right for Calico. You can't run multiple instances of Calico on a cluster, so the ability to select a namespace is substantially less useful for an application like Calico than it is for something like a web server or database, etc. 

I could imagine us allow configuration of the names of the namespaces we use, maybe, in order to match business conventions, but I still don't see a compelling reason to allow installation of Calico inside someone else's namespace.

**Comment by caseydavenport** [comment]
xref: https://github.com/projectcalico/calico/issues/5876

**Comment by caseydavenport** [comment]
As a more concrete example of why namespaces are important, see Pod Security Admission, which is an important replacement for Pod Security Policies: https://kubernetes.io/docs/concepts/security/pod-security-admission/

> Kubernetes offers a built-in Pod Security [admission controller](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/), the successor to [PodSecurityPolicies](https://kubernetes.io/docs/concepts/security/pod-security-policy/). Pod security restrictions are applied at the [namespace](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces) level when pods are created.

**Comment by michelzanini** [comment]
I am sorry but I still think you can't make decisions of how someone else's cluster will be setup.
You might design Calico into 3 namespaces as it might be a good for you, but other people might want to run it in just one.

If every add-on installed generated 3 namespaces, we would be in big trouble trying to find our workloads.

Also, people should be able to choose the name of the namespace they wish, if they have their own naming conventions.
Like I said, I don't see Helm charts out there that won't allow configuration of the namespace you want to run.

There's plenty of other missing configuration properties on Calico's Helm chart you usually find in any other chart.
Things like replica counts for the deployments, pod labels and annotations, resource limits, etc...
See this chart as example: https://github.com/stevehipwell/helm-charts/blob/master/charts/tigera-operator/values.yaml

**Comment by caseydavenport** [comment]
Resource limits were just added here: https://github.com/projectcalico/calico/pull/5934

I would gladly accept PRs that enable setting some of this other metadata - labels and annotations, etc., There are a lot of options that I 100% agree should be present from the link you shared - affinity, tolerations, nodeselector, etc., etc.

Looking at the linked values.yaml, HostNetwork is another example - if you allow configuration of `hostNetwork: false`, the operator app simply will not work. Maybe you could hack it (not sure why you'd want to), but this will largely just allow users to shoot themselves in the foot. We shouldn't enable that option just because someone else does. I lump namespaces in that bucket as well.

As an aside, if we did decide we want to enable installation of all of the Calico components into kube-system, there's a bit of work we'd need to do:

- The manifest -> operator migration logic uses the fact that there are separate namespaces in order to perform the migration.
- The operator itself uses Namespaces to allow creation of multiple resources with the same name, in some cases (this might just fall out, need to look further).

Neither of those are impossible to overcome - would just need some thinking and code. But it's extra work to enable what I still strongly believe is an anti-pattern and against good-practice.

>  we would be in big trouble trying to find our workloads

We could easily label the namespaces we create to make them easy to query as well (e.g., `kubectl get ns -l app=calico`). That seems reasonable.

My goal here isn't to dictate how your cluster should be set up. But I do have a goal of making it easy for our users to adhere to good security posture out of the box, and I would bet that any other Kubernetes maintainer will also tell you kube-system should be off-limits for user workloads. It's owned by the Kubernetes system itself and houses a number of privileged service accounts.

**Comment by caseydavenport** [comment]
Closing the loop on this one:

- Calico operator doesn't install in kube-system, and shouldn't. There are manifests available if you want to do that anyway.
- The helm chart includes ability to configure any of the options available in the Installation API, which includes replicas, resource requests and limits, annotations, topology configuration, etc.
- The helm chart supports a configurable namespace for the tigera-operator.
- There is ongoing work by @vara2504  to consolidate Calico down to a single namespace - `calico-system`. After that, perhaps we can look at making the name of that particular namespace configurable.

**Comment by RealOrko** [comment]
@caseydavenport I am with @michelzanini on this one. Your opinion is exactly that, your opinion. There are several reasons why housing platform tools in kube-system is preferred by platform teams. By your logic/reasoning we should probably tell AWS that hosting CSI/CNI extensions should be moved elsewhere too. Calico extends primitive constructs in kubernetes, so it most certainly is an extension. If most helm charts allow targetting a different namespace, is the internet wrong? It is a fairly simple ask with a simple answer. 

@michelzanini you should have led with: "I would like to host calico in the 'platform' namespace" ... would have spared you the long rant about k8s security and one persons opinion about how things should be organised in your cluster.

**Comment by caseydavenport** [comment]
> would have spared you the long rant about k8s security and one persons opinion

I'm afraid that being rude isn't going to win anyone any points here, and certainly not going to help motivate me as a core developer of this project to advocate for your use-case to the rest of the team over the many other enhancement requests we receive. I suggest re-reading this thread, and re-thinking your strategy.

Your characterization of this as an arbitrary personal opinion is pretty misguided. Especially since if you had read my last comment you would see that we were already working on consolidating into a single namespace and open to the idea of making that namespace configurable. I will quote it again for you below:

> There is ongoing work by vara2504 to consolidate Calico down to a single namespace - calico-system. After that, perhaps we can look at making the name of that particular namespace configurable.
